Automated Smoketest: :x:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-95.ca-central-1.compute.internal
Sha: 87998d85d8449ac1ae889c9559a69b991f72e919
Number of Error(s): 19
warn(226) A package without a profile could throw errors or warnings. package: dojo-util
warn(226) A package without a profile could throw errors or warnings. package: dojo-util
\terrors: 0
CMake Error at /home/centos/smoketest/PR-14872/build/generate-boost-regex-prefix/src/generate-boost-regex-stamp/download-generate-boost-regex.cmake:159 (message):
    error: downloading https://dl.bintray.com/boostorg/release/1.71.0/source/boost_1_71_0.tar.gz failed
  The requested URL returned error: 403 Forbidden
         error: downloading https://dl.bintray.com/boostorg/release/1.71.0/source/boost_1_71_0.tar.gz failed
  The requested URL returned error: 403 Forbidden
         error: downloading https://dl.bintray.com/boostorg/release/1.71.0/source/boost_1_71_0.tar.gz failed
  The requested URL returned error: 403 Forbidden
         error: downloading https://dl.bintray.com/boostorg/release/1.71.0/source/boost_1_71_0.tar.gz failed
  The requested URL returned error: 403 Forbidden
         error: downloading https://dl.bintray.com/boostorg/release/1.71.0/source/boost_1_71_0.tar.gz failed
  The requested URL returned error: 403 Forbidden
         error: downloading https://dl.bintray.com/boostorg/release/1.71.0/source/boost_1_71_0.tar.gz failed
  The requested URL returned error: 403 Forbidden
make[2]: *** [CMakeFiles/generate-boost-regex.dir/build.make:90: generate-boost-regex-prefix/src/generate-boost-regex-stamp/generate-boost-regex-download] Error 1
make[1]: *** [CMakeFiles/Makefile2:937: CMakeFiles/generate-boost-regex.dir/all] Error 2
make: *** [Makefile:163: all] Error 2
CMake Error at /home/centos/smoketest/PR-14872/build/generate-boost-regex-prefix/src/generate-boost-regex-stamp/download-generate-boost-regex.cmake:159 (message):
  Each download failed!
    error: downloading https://dl.bintray.com/boostorg/release/1.71.0/source/boost_1_71_0.tar.gz failed
         status_code: 22
         status_string: HTTP response code said error
         log:
         --- LOG BEGIN ---
           Trying 52.35.132.179...
  TCP_NODELAY set
  Connected to dl.bintray.com (52.35.132.179) port 443 (#0)
  ALPN, offering http/1.1
  successfully set certificate verify locations:
    CAfile: /etc/pki/tls/certs/ca-bundle.crt
    CApath: none
  [5 bytes data]
  TLSv1.3 (OUT), TLS handshake, Client hello (1):
  [512 bytes data]
  [5 bytes data]
  TLSv1.3 (IN), TLS handshake, Server hello (2):
  [102 bytes data]
  NPN, negotiated HTTP1.1
  [5 bytes data]
  TLSv1.2 (IN), TLS handshake, Certificate (11):
  [2765 bytes data]
  [5 bytes data]
  TLSv1.2 (IN), TLS handshake, Server key exchange (12):
  [333 bytes data]
  [5 bytes data]
  TLSv1.2 (IN), TLS handshake, Server finished (14):
  [4 bytes data]
  [5 bytes data]
  TLSv1.2 (OUT), TLS handshake, Client key exchange (16):
  [70 bytes data]
  [5 bytes data]
  TLSv1.2 (OUT), TLS change cipher, Change cipher spec (1):
  [1 bytes data]
  [5 bytes data]
  TLSv1.2 (OUT), TLS handshake, Next protocol (67):
  [36 bytes data]
  [5 bytes data]
  TLSv1.2 (OUT), TLS handshake, Finished (20):
  [16 bytes data]
  [5 bytes data]
  [5 bytes data]
  TLSv1.2 (IN), TLS handshake, Finished (20):
  [16 bytes data]
  SSL connection using TLSv1.2 / ECDHE-RSA-AES128-GCM-SHA256
  ALPN, server did not agree to a protocol
  Server certificate:
   subject: CN=*.bintray.com
   start date: Sep 26 00:00:00 2019 GMT
   expire date: Nov  9 12:00:00 2021 GMT
   subjectAltName: host dl.bintray.com matched cert\s *.bintray.com
   issuer: C=US; O=DigiCert Inc; OU=www.digicert.com; CN=GeoTrust RSA CA 2018
   SSL cer
 ... ( comment is too long, 1188 lines, 100599 bytes, truncated) ... 
[Build error:](https://gist.github.com/a3541fa5607a8a042a3e9ceb0873163e)
Build log [Build log](https://gist.github.com/HPCCSmoketest/a3541fa5607a8a042a3e9ceb0873163e#file-result-21-04-26-16-08-18-log)

```

--- a/dali/base/dadfs.cpp
+++ b/dali/base/dadfs.cpp
@@ -45,6 +45,7 @@
 #include <vector>
 #include <unordered_map>
 #include <algorithm>
+#include <time.h>
 
 #ifdef _DEBUG
 //#define EXTRA_LOGGING
@@ -174,6 +175,15 @@ const char *normalizeLFN(const char *s,StringBuffer &tmp)
 static IPropertyTree *getEmptyAttr()
 {
     return createPTree("Attr");
+}
+
+static double calcFileCost(const char * cluster, double sizeGB, double fileAgeDays)
+{
+    IPropertyTree * plane = queryStoragePlane(cluster);
+    if (!plane)
+        return 0.0;
+    double storageCostDaily = plane->getPropReal("cost/@storageAtRest", 0.0) * 12 / 365;
+    return storageCostDaily * sizeGB * fileAgeDays;
 }
 
 RemoteFilename &constructPartFilename(IGroup *grp,unsigned partno,unsigned partmax,const char *name,const char *partmask,const char *partdir,unsigned copy,ClusterPartDiskMapSpec &mspec,RemoteFilename &rfn)
@@ -4705,6 +4715,27 @@ public:
         else
             return false;
     }
+    virtual double getCost(const char * cluster) override
+    {
+        CDateTime dt;
+        getModificationTime(dt);
+        double fileAgeDays = difftime(time(nullptr), dt.getSimple())/(24*60*60);
+        double sizeGB = getDiskSize(true, false) / ((double)1024 * 1024 * 1024);
+
+        if (isEmptyString(cluster))
+        {
+            StringArray clusterNames;
+            unsigned countClusters = getClusterNames(clusterNames);
+            double totalCost = 0.0;
+            for (unsigned i = 0; i < countClusters; i++)
+                totalCost += calcFileCost(clusterNames[i], sizeGB, fileAgeDays);
+            return totalCost;
+        }
+        else
+        {
+            return calcFileCost(cluster, sizeGB, fileAgeDays);
+        }
+    }
 };
 
 static unsigned findSubFileOrd(const char *name)
@@ -6653,6 +6684,18 @@ public:
     virtual bool getSkewInfo(unsigned &maxSkew, unsigned &minSkew, unsigned &maxSkewPart, unsigned &minSkewPart, bool calculateIfMissing) override
     {
         return false;
+    }
+
+    virtual double getCost(const char * cluster) override
+    {
+        double totalCost = 0.0;
+        CriticalBlock block (sect);
+        ForEachItemIn(i,subfiles)
+        {
+            IDistributedFile &f = subfiles.item(i);
+            totalCost += f.getCost(cluster);
+        }
+        return totalCost;
     }
 };
 

--- a/dali/base/dadfs.hpp
+++ b/dali/base/dadfs.hpp
@@ -415,6 +415,7 @@ interface IDistributedFile: extends IInterface
     virtual bool getSkewInfo(unsigned &maxSkew, unsigned &minSkew, unsigned &maxSkewPart, unsigned &minSkewPart, bool calculateIfMissing) = 0;
     virtual int  getExpire() = 0;
     virtual void setExpire(int expireDays) = 0;
+    virtual double getCost(const char * cluster) = 0;
 };
 
 

--- a/esp/scm/ws_dfu.ecm
+++ b/esp/scm/ws_dfu.ecm
@@ -191,6 +191,7 @@ ESPStruct [nil_remove] DFUFileDetail
     [min_ver("1.39")] ESParray<ESPstruct DFUFileBloom> Blooms;
     [min_ver("1.40")] int ExpireDays;
     [min_ver("1.41")] string KeyType;
+    [min_ver("1.59")] string Cost;
 };
 
 ESPStruct DFUSpaceItem
@@ -967,8 +968,8 @@ ESPresponse [exceptions_inline, nil_remove] DFUFilePublishResponse
 //  ===========================================================================
 ESPservice [
     auth_feature("DEFERRED"),
-    version("1.58"),
-    default_client_version("1.58"),
+    version("1.59"),
+    default_client_version("1.59"),
     noforms,
     exceptions_inline("./smc_xslt/exceptions.xslt")] WsDfu
 {

--- a/esp/services/ws_dfu/ws_dfuService.cpp
+++ b/esp/services/ws_dfu/ws_dfuService.cpp
@@ -45,6 +45,7 @@
 #include "fverror.hpp"
 #include "nbcd.hpp"
 #include "thorcommon.hpp"
+#include "jstats.h"
 
 #include "jstring.hpp"
 #include "exception_util.hpp"
@@ -2643,6 +2644,13 @@ void CWsDfuEx::doGetFileDetails(IEspContext &context, IUserDescriptor *udesc, co
             if (includeJsonTypeInfo)
                 FileDetails.setJsonInfo(jsonLayout);
         }
+    }
+    if (version >= 1.59)
+    {
+        double totalCost = df->getCost(cluster);
+        StringBuffer s;
+        formatMoney(s, money2cost_type(totalCost));
+        FileDetails.setCost(s);
     }
     PROGLOG("doGetFileDetails: %s done", name);
 }

--- a/helm/hpcc/values.schema.json
+++ b/helm/hpcc/values.schema.json
@@ -424,9 +424,20 @@
           "type" : "string"
         },
         "defaultSprayParts" : {
-            "description": "Number of parts sprayed by default",
-            "type" : "integer"
+          "description": "Number of parts sprayed by default",
+          "type" : "integer"
+        },
+        "cost" : {
+          "description": "Costs associated with the storage and use of the plane",
+          "type" : "object",
+          "properties": {
+            "storageAtRest" : {
+              "description": "Storage cost (GiB/month)",
+              "type": "number"
+            }
           }
+        }
+
       },
       "required": [ "name", "prefix" ],
       "additionalProperties": false

--- a/helm/hpcc/values.yaml
+++ b/helm/hpcc/values.yaml
@@ -96,6 +96,8 @@ storage:
   #   hosts: <name>                         # Name of the host group for bare metal - must match the name of the storage plane..
   #   secret: <secret-id>                   # what secret is required to access the files.  This could optionally become a list if required (or add secrets:).
   #   defaultSprayParts: 4                  # The number of partitions created when spraying (default: 1)
+  #   cost:                                 # The storage cost
+  #     storageAtRest: 0.113                # Storage at rest cost: cost per GiB/month
   #   options:                              # not sure if it is needed
 
   dllStorage:

--- a/system/jlib/jstats.cpp
+++ b/system/jlib/jstats.cpp
@@ -438,7 +438,7 @@ private:
 static MoneyLocale moneyLocale;
 CriticalSection MoneyLocale::cslock;
 
-static StringBuffer & formatMoney(StringBuffer &out, unsigned __int64 value)
+StringBuffer & formatMoney(StringBuffer &out, unsigned __int64 value)
 {
     std::stringstream ss;
     std::locale & loc = moneyLocale.queryMoneyLocale();

--- a/system/jlib/jstats.h
+++ b/system/jlib/jstats.h
@@ -824,5 +824,6 @@ public:
     }
 };
 
+extern jlib_decl StringBuffer & formatMoney(StringBuffer &out, unsigned __int64 value);
 
 #endif

